### PR TITLE
changes to add a few more features

### PR DESCRIPTION
**Pull request by snagytx**
these changes are based on develop tree

1) add retry logic
2) add an 520 error text
3) return some extra fields in the to_info call

### CODE REVIEW

**Comment by Razzeee** [comment]
@snagytx 
seems to be some kind of to_mediaObject not exactly to_Info?
and to_Info for show is missing rating.

**Comment by snagytx** [comment]
both are true.

My view on to_info is to return the data for that object and it's children
rating can be added for show object if needed, I just didn't need it in my work

**Comment by Razzeee** [comment]
Then we need @fuzeman to know what he is using it for in trakt for plex
I would only add the ratings for consistency reasons.

**Comment by snagytx** [comment]
I'm OK with adding a new function that would do what I added to to_info if that's needed

**Comment by snagytx** [comment]
added ratings in this pull

**Comment by Razzeee** [comment]
Let's wait for @fuzeman then
Does movies have everything it needs (when we're already at it)

**Comment by Razzeee** [comment]
If this gets added it would be nice for movies to be similar, like:

    def to_info(self):
        return {
            'ids': dict(self.keys),
            'title': self.title,
            'year': self.year,
            'watched': 1 if self.is_watched else 0,
            'collected': 1 if self.is_collected else 0,
            'plays': self.plays,
            'rating': self.rating,
            'collected_at': self.collected_at
        }

**Comment by Razzeee** [comment]
@snagytx 
just found out the hard way, rating should be self.rating.value everywhere otherwise we'll get the object and thats breaking the parsing

**Comment by Razzeee** [comment]
@snagytx 
please check https://github.com/Razzeee/script.trakt/commit/0b2ccaa461a4154dd607619fc9c49df13d05b042
for reference

**Comment by fuzeman** [comment]
I'm currently using ```to_info()``` to return an identifier of a movie/show (for use in /sync methods).

So, I'm planning on copying the original ```to_info()``` method and naming it ```to_identifier()```, then have ```to_info()``` actually return all the data for the object (the changes from this PR).

-------------------------

I'm not sure how trakt actually wants us to handle 5xx errors, because having all the XBMC & Plex clients continuing to hit the server when they encounter 5xx errors might not be a good idea...

I'll merge request retrying anyway, but will include a bit more configuration via ```Trakt.configuration``` (e.g. enable/disable, maximum retries)

**Comment by Razzeee** [comment]
@snagytx 
Please add the rating changes from my commit above  to this pull request, thx

**Comment by Razzeee** [comment]
@fuzeman 
to_identifier seems to be more fitting. 
But I'm not sure about the other one being to_info. Their named mediaObjects in the trakt docs.

**Comment by fuzeman** [comment]
Finished merging this PR into the develop branch. I've also included the fix for ratings in to_info() (as @Razzeee suggested).

I'm not sure about the name of ```to_info()```, they aren't exactly trakt "standard media objects" (the structure is a bit different).. maybe ```to_dict()``` would be a better name?

--------------

I've decided to disable retrying by default, but it can be enabled for all requests with:
```python
Trakt.configuration.defaults.http(retry=True)
```

or for specific requests with:
```python
with Trakt.configuration.http(retry=True):
    watched_shows = Trakt['sync/watched'].shows()
```

**Comment by Razzeee** [comment]
@fuzeman 
Are they really different? I'm using em to create the mediaObjects for my calls. Sure, there are times, where I want to delete collected_at first before sending it off.

**Comment by fuzeman** [comment]
```to_info()``` returns a ```dict``` representation of the trakt.py object - keys like "watched", "collected" and "plays" aren't accepted anywhere in the trakt.tv API.

You would need to filter out some of the keys for each API call otherwise you are just sending extra data for no reason *(e.g. the remove() methods only need the media identifier, not "rating", "plays", etc...)*.

**Comment by Razzeee** [comment]
@fuzeman 
your right.
how are you handling these in plex trakt? (mediaObject creation for querys etc)

**Comment by fuzeman** [comment]
Bit hard to explain, but I'll try..

Basically for each episode and movie in Plex a [watch()](https://github.com/trakt/Plex-Trakt-Scrobbler/blob/f1e24fea59685a3c61a0e05081d0e339e98e2289/Trakttv.bundle/Contents/Code/sync/push.py#L18-51), [rate()](https://github.com/trakt/Plex-Trakt-Scrobbler/blob/f1e24fea59685a3c61a0e05081d0e339e98e2289/Trakttv.bundle/Contents/Code/sync/push.py#L53-88) and [collect()](https://github.com/trakt/Plex-Trakt-Scrobbler/blob/f1e24fea59685a3c61a0e05081d0e339e98e2289/Trakttv.bundle/Contents/Code/sync/push.py#L90-114) function is run - if an update to trakt.tv is required an `item` is built (this includes the identifier ```title, year, ids``` and any extra data ```collected_at, watched_at, rating```), they are then collected in a list ready to be [sent to trakt.tv](https://github.com/trakt/Plex-Trakt-Scrobbler/blob/f1e24fea59685a3c61a0e05081d0e339e98e2289/Trakttv.bundle/Contents/Code/sync/push.py#L390-394).

The linked functions will give you more of an idea how it's implemented, it might be a little hard to follow how the functions are called though.

**Comment by Razzeee** [comment]
Okay, so your building it on demand from your plex data with to_trakt from your plex implementation and some more fields that are different for all requests.

**Comment by Razzeee** [comment]
@fuzeman 
just to clarify this code

```
Trakt.configuration.http(retry=True, max_retries=1)`
```

will lead to two executions if there is an error? one normal and one retry?

**Comment by fuzeman** [comment]
@Razzeee yes *(https://github.com/fuzeman/trakt.py/blob/develop/trakt/core/http.py#L52)*
